### ISSUE TITLE
Identifying strings are expressions in place where the string might also be a constant

### ISSUE THREAD

**Issue by fjtirado** [serverlessworkflow/specification#848]
Currently we can write
```
set: 
  name: Javi

```
or 
```
set: 
   name: ${.input.name}
```
However $ is a reserved keyword in JQ and {} are used in json, think of this expression written in json
```
"set": {"name": "${$loopVariable}"} 
```
(I wrote it wrong three times :))

I think we should consider following options

1. Everything is an expression

This implies user has to escape literals. 

```
set: 
  name: '"Javi"'
```
```
set: 
   name: .input.name
```

2. Change the expression  prefix. 

Use something different than `${}` for example, use ``

```
set: 
  name: Javi
```
```
set: 
   name: `.input.name`
```

We can use a different character, for example, `#`, but with this one we need to add white spaces around the expression
```
set: 
   name: # .input.name #
```

3. Use a new keyword. 

So after the property name, you expect either `constant` or `expr` and use as default `expr` (that default might be overriden by user in the header of the workflow definition)

- Default to `expr`:

```
set: 
  name:
    constant: Javi
```
```
set: 
   name: .input.name
```
- Default to `constant`: 

```
set: 
  name:  Javi
```
```
set: 
   name: 
     expr:  .input.name
```

**Comment by gibson042** [comment]
Relevant prior discussion: #285 (https://github.com/serverlessworkflow/specification/issues/285#issuecomment-853431146 even includes several of these options)

**Comment by cdavernas** [comment]
As discussed in PM, I totally understand your point of view, which is shared insightfully by others in the issue shared by @gibson042.

As I have never been bothered by the current approach, why i actually like, I'm probably not the most objective person when it comes to that subject.

I did, however, my due diligence, and have checked what major clouds are proposing, and it seems they are all using a similar mechanism. Hell, Google Workflows, which is my personnal favorite,  uses the exact same than us. 

They, however, recommend single quoting expressions in YAML (which is/should be the authoring language) to avoid many of the problems you guys mentionned.

**Comment by fjtirado** [comment]
@cdavernas 
Google Expressions does not use JQ, thats why starting with $ is not an issue. 
The model is not a json structure, thats why using {} is not an issue either. 
Maybe it was lost in my long description but this expression in json looks really confusing when defining the workflow in json. 
`"set": {"name": "${$loopVariable}"} `
We at least need a different enclosing set of characters.

**Comment by cdavernas** [comment]
> when defining the workflow in json.

Well, why would you go through the sheer pain of defining anything in JSON when you can do it in YAML, I wonder?

**Comment by fjtirado** [comment]
@cdavernas 
Believe or not, for me is easier to use {} (json) than tabs (yaml)  (I am suspect that for other developers that consider that python practise a bad joke)
Regardless my preferences, since the definition file can still be written in json and even in yaml you might want to write expressions that return a json, for example 
``set: '${{loopVariable:$loopVariable}}'`
I am really surprised thay you consider ${} a good separator for a definition using JQ and Json data everywhere just because it was done that way before (specially when we are defining a completely new DSL)

**Comment by cdavernas** [comment]
> Believe or not, for me is easier to use {} (json) than tabs (yaml) 

I believe you, and that's your right. But I am convinced that it will probably not be the case of the majority of our (especially non-technical) audience, which I'd rather focus on (as per design).

> I am really surprised thay you consider ${} a good separator for a specification using JQ and Json data everywhere just because it was done that way before (specially when we are defining a completely new DSL)

Well, first, it's a matter of taste and colors, really. Second, I was looking to what was done where and why. The reasons identified and addressed by our "competitors" is of high interest to me.

The fact is that while having to build a JSON string should be doable, it's rather something you'd like to avoid, as you'd loose control over the data, which becomes opaque to any component of you workflow.

Please rest assured I read through and through the discussion we had on slack, the issue #285 and the new example supplied above. My opinion/taste remains the same nonetheless. 

I therefore propose we all avoid trying to convince each other, and rather provide constructive solutions that could address the issue at hand!

I still believe we are speaking of edge cases twhich can be addressed in a way that becomes painfull for standard cases.

Please, let's remain concise in this issue, and move discussions to... discussions. I propose we only make and discuss proposals in here!

**Comment by fjtirado** [comment]
Given your personal preferences, my proposal is to change `${}` by` ##` (which is more neutral in jq and json)

**Comment by cdavernas** [comment]
@ricardozanini @JBBianchi @gibson042 @jorgenj WDYT?

**Comment by fjtirado** [comment]
In https://github.com/serverlessworkflow/specification/issues/285 I think we were both discussing (makes sense at that moment)  expression interpolation (which we are not supporting anymore)  with expression identification (and force the use of ${} in place were it was not really needed because it was obvious the string was expression, problem that is not there anymore)
The only problem that remains is to identify expression in places where it is not evident that the string is an expression or a literal. So far, in the new DSL, I believe that will only occur in few places (set: and body:).

**Comment by ricardozanini** [comment]
1. +1 for the proposal to set `implicit | explicit` at the beginning of the workflow definition. We might explore the usage of `expressions: strict` or `strict: true` since it's widespread among many languages. Again, a matter of choice since we all agree that we need this feature.
2. +1 to change from `${}` to `$()` to avoid misinterpretation with JSON expressions. Although one can argue that `()` is super common in expressions that might also bring misinformation to the expression. From jq manual: `Parenthesis work as a grouping operator just as in any typical programming language.` Again, a matter of choice.
3. Welcome back @gibson042, please stay around

**Comment by fjtirado** [comment]
@ricardozanini what we are doing is writing an string that contains an expression that is evaluated. 
 () can be understood as synonomous to evaluate/call and $ might be understood as  a shortcut for expression.

**Comment by ricardozanini** [comment]
@fjtirado I'm just adding to the discussion, I already agreed with `()`.

**Comment by fjtirado** [comment]
Also, we should be more strict defining what the expression should be returning. 
In the cases we are discussing, set: and body: the expression is returning any valid json node (object, arrary or literal)
In the case of switch, the expression should return a boolean literal.
In the case of username for basis authorization, the expressions should return a string literal. 
In the case of any timeout, the expression should return a number literal.
And so on.

**Comment by fjtirado** [comment]
Ok, since it seems we have an agreement once @cdavernas is done with the DSL refactor, I can will open a PR associated to this issue modifying the spec and the examples.

**Comment by ricardozanini** [comment]
@fjtirado I'll introduce examples + CI here: #852 #850.

**Comment by cdavernas** [comment]
> Ok, it seems we have an agreement. Once @cdavernas is done with the DSL refactor, I will open a PR associated to this issue modifying the spec and the examples.

I'm not necessarily for that change, I don't think that using '$( )' is less confusing than '${ }' or causes less problems, especially with the `loose` mode added.

Plus, like discussed with @ricardozanini, I'd like before we make any decision to get @JBBianchi's input, which will be possible next week or so.

Ideally, we should gather as much "opinions" on this to see which one "wins", if any. Maybe setting a small poll?

**Comment by fjtirado** [comment]
@cdavernas It not a question of personal choice. Using {} as markup character in json is confusing. I think the example I put is pretty clear.

**Comment by cdavernas** [comment]
AFAIK, in YAML you do not have to do anything, because happilly it starts with ${, not simply {.

The reasoning behind using mustaches, which people can understandingly be nervous about, is because it's conventionnally used to indicate evaluation.

I'm not saying we should/should not use it, - hell, I like your suggestion-, I'm just saying I would like other people opinion before claiming unilaterally that we have an agreement 😛 

Can we just wait for start of next week before taking a decision/acting?

**Comment by fjtirado** [comment]
I agree we can wait till next week, I was raising the point becuase there is another PR where the ${} appeared (the looss evaluation one) and we need to wait to have an agreement before that PR was merged

**Comment by cdavernas** [comment]
It was a quick addition to the refactor, to enable the "loose" use case ;)

**Comment by cdavernas** [comment]
> In YAML you need to either use white spaces or quoutes, since { and } are special characters that needs to be esscaped

This is valid, with no quotes, no whitespaces:

```yaml
test:
  expression: ${ Hello, world }
```

```yaml
test:
  expression: ${Hello, world}
```

Am I missing something?

**Comment by cdavernas** [comment]
> I agree we can wait till next week, I was raising the point becuase there is another PR where the ${} appeared (the looss evaluation one) and we need to wait to have an agreement before that PR is merged

That PR already was merged, as part of the refactor! I added the loose/strict before pushing it in, based on your/ric's awesome suggestions!

**Comment by fjtirado** [comment]
```
test:
  expression: ${ Hello, world }
```
should be valid  because the special character {  and }  are separated by whitespaces

```
test:
  expression: ${Hello, world}
```
should not be (if it is, maybe the parser is smart and intepreting that {} are not special characted but part of the string because the string does not start with {. 

This is the yaml spec section I was using https://yaml.org/spec/1.2.2/#742-flow-mappings

**Comment by cdavernas** [comment]
> should not be (if it is, maybe the parser you are using is smart and intepreting that {} are not special characted but part of the string because the string does not start with {, but my point still stands

I dont want to be a PITA, especially that, once again, I'm not opposed to the change, but I disagree when you say your point still stands.

As a matter of fact, YAML doesnt tell you you shouldnt be using mustaches, it just tells you that if you encounter this token while reading, and before the token has been identified as, say, a string, it should denotate the start of a mapping.

If the line starts with, say, $, the token IS a string, and YAML allows you to write whatever you want, including mustaches, semi-colons, dots, which all are other control characters.

That fact goes against your point, saying that we can't use { because it's a control character. We are not using '{', we are using '${', which simply works!

On a side note to that, I want to remind you mustaches are what Amazon and Google have opted for. I'm not saying, once again, that we should use them, or that this fact is a good reason to use them, I just think that if YAML intended such limitations (which it does not, every linter I tried works without quotes), high chances are they would have chosen other characters.

**Comment by fjtirado** [comment]
if yaml is so smart and you can write everything after the $, then why you write the swith example using quotes
` when: '.color == "red"'`
If if starts with a ., it should be a string, isnt it?
But I would be happy if you point me to the specification section of yaml when thats stated so we can be sure that ${ working without qoutes is not a side effect of a particular yaml parser understanding but an universal fact. 
Anyway, I think the yaml discussion is pointles, event if ${ is guaranteed to work, { and } are a visual problem when writing in json, as I tried to explain (apparently unsucessfully)

**Comment by fjtirado** [comment]
And now, lets discuss about google workflows
This is a snipper of one of their examples
```
body:
        kind: "sql#database"
        charset: "utf8"
        collation: "utf8_general_ci"
        name: ${database_name}
        instance: ${instance}
        project: ${project}
        sqlserverDatabaseDetails:
          compatibilityLevel: 5
          recoveryModel: "Simple"
```
They are using ${} for their expression without qoutes (so it seems ignoring { and } in indeed universal, as you claimed, my bad, I will review the yaml spec to check where that is stated) but they are significantly wrapping constants within ""
In their expressions, they are not using json constructs (thats why { is not going to be ever an issue for them) and they are forcing the qoutes (so they are fine with escaping qoutes in json)

**Comment by fjtirado** [comment]
@cdavernas  Btw, I do not think we are being a PITA. When writing a spec we need to be rigorous. I found this conversation enriching

**Comment by cdavernas** [comment]
> But I would be happy if you point me to the specification section of yaml when thats stated so we can be sure that ${ working without qoutes is not a side effect of a particular yaml parser understanding but an universal fact.

I cannot do that, unhappilly, nor could you do the opposite. I'm however open to conceive an interpreter or another might not handle the use case you presented properly. Could you find one that didnt like the ${ ?

> { and } are a visual problem when writing in json

That is a very valid point, and is probably the essence of your proposal. 
To play the devil's advocate, one could however argue that a, it is a matter of perception, b, that you should be writting in YAML, not JSON, but that also is a matter of personnal preference.

> Btw, I do not think we are being a PITA. When writing a spec we need to be rigorous. I found this conversation enriching

Cool! And so do I!

**Comment by fjtirado** [comment]
I was playing a bit with the current conversion between YAML and JSON that I wrote using Jackson. 

YAML snippet 

```
       mantra: Serverless Workflow is awesome!
    boolean: true
    number: 2
    json: {"pepe":2}
    jqexpr: ${{pepe:2}} #jq expr that returns json object {"pepe":2}
    jqexprusingdifferentmarkup: .{{pepe:2}}
```

is converted to this JSON
```
       "mantra" : "Serverless Workflow is awesome!",
      "boolean" : true,
      "number" : 2,
      "json" : {
        "pepe" : 2
      },
      "jqexpr" : "${{pepe:2}}",
      "jqexprusingdifferentmarkup" : ".{{pepe:2}}"

```
This confirms that very likely any yaml parser which found any character before the { will understand that { is not declaring a json

**Comment by fjtirado** [comment]
Also
YAML
`when: .color == "red"`
JSON
` "when" : ".color == \"red\""`
@cdavernas  So quoutes are not needed to wrap when condition

**Comment by cdavernas** [comment]
@fjtirado do we still need to do anything here, even after adding the `evaluate` top level property and related options?
